src/instructlab/configuration.py

```diff
@@ -17,55 +18,190 @@
-DEFAULT_PROMPT_FILE = "prompt.txt"
-DEFAULT_GENERATED_FILES_OUTPUT_DIR = "generated"
-DEFAULT_CONNECTION_TIMEOUT = httpx.Timeout(timeout=30.0)
-# use spawn start method, fork is not thread-safe
-DEFAULT_MULTIPROCESSING_START_METHOD = "spawn"
-
-# When otherwise unknown, ilab uses this as the default family
-DEFAULT_MODEL_FAMILY = "merlinite"
+ILAB_PACKAGE_NAME = "instructlab"
+CONFIG_FILENAME = "config.yaml"
+
+
+class STORAGE_DIR_NAMES:
+    ILAB = "instructlab"
+    DATASETS = "datasets"
+    CHECKPOINTS = "checkpoints"
+    MODELS = "models"
+    TAXONOMY = "taxonomy"
+    INTERNAL = (
+        "internal"  # for storing all ilab-internal files the user doesn't need to see
+    )
+    CHATLOGS = "chatlogs"
+
+
+class _InstructlabDefaults:
+    """
+    Class that defines the default paths used by ilab.
+    We define them this way so that they can be lazy-loaded and overridden by tests.
+    This way, they are defined when they are read instead of when the module is imported.
+    """
+
+    # define static defaults up here
+    API_KEY = "no_api_key"
+
+    # TODO: Consolidate --model and --model-path into one --model-path flag since we always need a path now
+    MODEL_NAME = "merlinite-7b-lab-Q4_K_M"
+    MERLINITE_GGUF_REPO = "instructlab/merlinite-7b-lab-GGUF"
+    GGUF_MODEL_NAME = "merlinite-7b-lab-Q4_K_M.gguf"
+    MODEL_REPO = "instructlab/granite-7b-lab"
+    JUDGE_MODEL_MT = "prometheus-eval/prometheus-8x7b-v2.0"
+    TAXONOMY_REPO = "https://github.com/instructlab/taxonomy.git"
+    TAXONOMY_BASE = "origin/main"
+    MAX_CONTEXT_SIZE = 4096
+    # TODO: these constants should be removed, they should not leak out
+    NUM_CPUS = 10
+    CHUNK_WORD_COUNT = 1000
+    NUM_INSTRUCTIONS = 100
+    CONNECTION_TIMEOUT = httpx.Timeout(timeout=30.0)
+    # use spawn start method, fork is not thread-safe
+    MULTIPROCESSING_START_METHOD = "spawn"
+
+    # When otherwise unknown, ilab uses this as the default family
+    MODEL_FAMILY = "merlinite"
+
+    def __init__(self):
+        self._cache = {}
+
+    def _clear_cache(self):
+        """
+        Utility function which is mostly used for testing purposes to clear the cache.
+        Otherwise, all tests will used cached temporary directories and cause errors.
+        """
+        self._cache = {}
+
+    def _set_and_retrieve(self, key: str, value: str) -> str:
+        """
+        Retrieves value associated with the given key. If it's not present in the cache,
+        the provided value is stored in the cache and returned.
+        """
+        if key not in self._cache:
+            self._cache[key] = value
+        v: str | typing.Any = self._cache[key]
+        if not isinstance(v, str):
+            raise ValueError(f"Expected {key=} to be a string, got {v=}: {type(v)}")
+        return v
+
+    @property
+    def CONFIG_HOME(self) -> str:
+        return self._set_and_retrieve(
+            "CONFIG_HOME", platformdirs.user_config_dir(ILAB_PACKAGE_NAME)
+        )
```

Comment:
I didn't notice this code before, because GH had collapsed the diff for `configuration.py`.

The way how you wrote this code, will make it slower, not faster. Every propery access will evaluate the default (2nd parameter) and then do extra work to look up the key in a dict. I recommend that you simplify this code and remove all caching. Path lookup is not a performance critical part of InstructLab. Besides it is super faster already. On my system, platformdirs lookup + `os.path.join` takes less than 5 microseconds:

```
$ python3 -m timeit -s 'import os, platformdirs' 'os.path.join(platformdirs.user_data_dir("instructlab"), "checkpoints")'
50000 loops, best of 5: 4.54 usec per loop
```

```suggestion
    @property
    def CONFIG_HOME(self) -> str:
        return platformdirs.user_config_dir(ILAB_PACKAGE_NAME)
```

Reply:
@tiran The point of caching this is not to provide a performance speedup but instead to prevent these values from being defined until they are needed.

The reason to cache here is so that after they are loaded, the variable (and all other common dependencies) consistently provides the same answer, even if the underlying envs change further in the script.

As to the point about speed, I wouldn't worry about it as these are O(1) operations. This is Python afterall, so things will only be so fast.

Reply:
The only thing that could change behavior are env vars like HOME or XDG_DATA_HOME. If consistency is your concern, then you can cache the value of the three directories from platformdirs early. This allows for consistent values and still allows tests to `mock.patch` values:

```python
class _InstructlabDefaults:
    _cache_dir: str
    _config_dir: str
    _data_dir: str

    def __init__(self):
        self._reset()
    
    def _reset(self):
        pd = platformdirs.PlatformDirs(appname=ILAB_PACKAGE_NAME)
        self._cache_home = pd.user_cache_dir
        self._config_dir = pd.user_config_dir
        self._data_dir = pd.user_data_dir
    
    @property
    def CHECKPOINTS(self) -> str:
        return path.join(self._cache_home, STORAGE_DIR_NAMES.CHECKPOINTS)
```

Reply:
Thanks @tiran, I'll use this method then.

Reply:
This is much cleaner!